plugins/uefi/fu-plugin-uefi.c

```diff
@@ -135,6 +137,202 @@ fu_plugin_get_results (FuPlugin *plugin, FuDevice *device, GError **error)
+			     tmp, localedir);
+		return NULL;
+	}
+
+	/* decompress data */
+	stream_compressed = g_memory_input_stream_new_from_bytes (compressed_data);
+	conv = G_CONVERTER (g_zlib_decompressor_new (G_ZLIB_COMPRESSOR_FORMAT_GZIP));
+	stream_raw = g_converter_input_stream_new (stream_compressed, conv);
+	buf = g_malloc0 (buf_sz);
+	while ((len = g_input_stream_read (stream_raw,
+					   buf + buf_idx,
+					   buf_sz - buf_idx,
+					   NULL, error)) > 0) {
+		buf_idx += len;
+		if (buf_sz - buf_idx < chunk_size) {
+			buf_sz += chunk_size;
+			buf = g_realloc (buf, buf_sz);
+		}
+	}
+	if (len < 0) {
+		g_prefix_error (error, "failed to decompress file: ");
+		return NULL;
+	}
+	g_debug ("decompressed image to %" G_GSIZE_FORMAT "kb", buf_idx / 1024);
+	return g_bytes_new_take (g_steal_pointer (&buf), buf_idx);
+}
+
+static gboolean
+fu_plugin_uefi_update_splash (GError **error)
+{
+	fwup_resource *re = NULL;
+	guint best_idx = G_MAXUINT;
+	guint32 lowest_border_pixels = G_MAXUINT;
+#ifdef HAVE_FWUPD_GET_BGRT_INFO
+	int rc;
+#endif
+	guint32 screen_height = 640;
+	guint32 screen_width = 480;
+	g_autoptr(fwup_resource_iter) iter = NULL;
+	g_autoptr(GBytes) image_bmp = NULL;
+	struct {
+		guint32	 width;
+		guint32	 height;
+	} sizes[] = {
+		{ 640, 480 },
+		{ 800, 600 },
+		{ 1024, 768 },
+		{ 1920, 1080 },
+		{ 3840, 2160 },
+		{ 5120, 2880 },
+		{ 5688, 3200 },
+		{ 7680, 4320 },
+		{ 0, 0 }
+	};
+
+	/* is this supported? */
+	fwup_resource_iter_create (&iter);
+	re = fu_plugin_uefi_find (iter, UEFI_GUID_IMAGE_UPLOAD, error);
+	if (re == NULL)
+		return TRUE;
+
+	/* get the boot graphics resource table data */
+#ifdef HAVE_FWUPD_GET_BGRT_INFO
+	rc = fwup_get_ux_capsule_info (&screen_width, &screen_height);
+	if (rc < 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "failed to get BGRT screen size");
+		return FALSE;
+	}
+	g_debug ("BGRT screen size %" G_GUINT32_FORMAT " x%" G_GUINT32_FORMAT,
+		 screen_width, screen_height);
+#endif
+
+	/* find the 'best sized' pre-generated image */
+	for (guint i = 0; sizes[i].width != 0; i++) {
+		guint border_pixels;
+
+		/* disregard any images that are bigger than the screen */
+		if (sizes[i].width > screen_width)
+			continue;
+		if (sizes[i].height > screen_height)
+			continue;
+
+		/* is this the best fit for the display */
+		border_pixels = (screen_width * screen_height) -
+				(sizes[i].width * sizes[i].height);
+		if (border_pixels < lowest_border_pixels) {
+			lowest_border_pixels = border_pixels;
+			best_idx = i;
+		}
+	}
+	if (best_idx == G_MAXUINT) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "failed to find a suitable image to use");
+		return FALSE;
+	}
+
+	/* get the raw data */
+	image_bmp = fu_plugin_uefi_get_splash_data (sizes[best_idx].width,
+						    sizes[best_idx].height,
+						    error);
+	if (image_bmp == NULL)
+		return FALSE;
```

Comment:
Of course if you do end up passing a pointer (per my other comment) this would need to be adjusted.